docs/async/async-csharp-vb.rst

```diff
@@ -0,0 +1,213 @@
+
+* ``async void`` **should only be used for event handlers.**
+
+Why?  It's the only way to allow asynchronous event handlers work because events do not have return types (thus cannot make use of ``Task`` and ``Task<T>``).  Any other use of ``async void``does not follow the Task-based model and will bring about some issues, such as:
+
+    (a) Exceptions thrown in an ``async void`` method can't be caught outside of that method.
+
+    (b) ``async void`` methods are very difficult to test.
+
+    (c) ``async void`` methods can cause bad side effects if the caller isn't expecting them to be async.
+
+* **Tread carefully when using async lambdas in LINQ expressions**
+
+Lambda expressions in LINQ use deferred execution, meaning code could end up executing at a time when you're not expecting it to.  The introduction of blocking tasks into this can easily result in a deadlock if not written correctly.  The nesting of asynchronous code like this can also make it more difficult to reason about the execution of the code.  Async and LINQ are powerful, but if they are abused they can make things difficult.  Clarity is always better than cleverness.
+
+* **Write code that awaits Tasks in a non-blocking manner**
+
+Blocking the current thread as a means to wait for a Task to complete can result in deadlocks, blocked context threads, and significantly more complex error-handling.  The following table should provide some guidance in how to deal with waiting for Tasks in a non-blocking way:
+
+====================== ================================= =======================
+Use this...            Instead of this...                When wishing to do this
+====================== ================================= =======================
+``await``              ``Task.Wait`` or ``Task.Result``  Retrieving the result of a background task
+``await Task.WhenAny`` ``Task.WaitAny``                  Waiting for any task to complete
+``await Task.WhenAll`` ``Task.WaitAll``                  Waiting for all tasks to complete
+``await Task.Delay``   ``Thread.Sleep``                  Waiting for a period of time
+====================== ================================= =======================
+
+* **Write less stateful code**
+
+Don't depend on the state of global objects or the execution of certain methods.  Instead, depend only on the return values of methods.  Why?
+
+	(a) Code will be easier to reason about
+	(b) Code will be easier to test
+	(c) Mixing async and synchronous code is far simpler
+	(d) Race conditions can typically be avoided altogether
+	(e) Depending on return values makes coordinating async code simple
+	(f) (Bonus) it works really well with dependency injection
+
+For a specific goal, aim for complete or near-complete `Referential Transparency <https://en.wikipedia.org/wiki/Referential_transparency_(computer_science)>`_ in your code.  Doing so will result in an extremely predictable, testable, and maintainable codebase.
```

Comment:
"For a specific goal" seems like an odd statement. I don't know how to think of that.